vowpalwabbit/interactions_predict.h

```diff
@@ -205,46 +202,38 @@ inline void generate_interactions(const std::vector<std::vector<namespace_index>
 
 #endif
     {
+      state_data.clear();
       bool must_skip_interaction = false;
       // preparing state data
-      feature_gen_data* fgd = state_data.begin();
-      feature_gen_data* fgd2;  // for further use
       for (auto n : ns)
       {
-        features& ft = features_data[static_cast<int32_t>(n)];
-        const size_t ft_cnt = ft.indicies.size();
-
+        const auto& ft = features_data[static_cast<size_t>(n)];
+        const auto ft_cnt = ft.size();
         if (ft_cnt == 0)
         {
           must_skip_interaction = true;
           break;
         }
-
-        if (fgd == state_data.end())
-        {
-          state_data.push_back(empty_ns_data);
-          fgd = state_data.end() - 1;  // reassign as memory could be realloced
-        }
-
-        fgd->loop_end = ft_cnt - 1;  // saving number of features for each namespace
-        fgd->ft_arr = &ft;
-        ++fgd;
+        state_data.push_back(feature_gen_data{ft.audit_begin(), ft.audit_end()});
       }
 
       // if any of interacting namespace has 0 features - whole interaction is skipped
       if (must_skip_interaction) continue;  // no_data_to_interact
 
+      auto fgd = state_data.data();                             // always equal to first ns
```

Comment:
nit: might be a little easier to read with
```
auto fgd = &state_data.front();
auto fgd2 = &state_data.back();
```

Doesn't really make a difference though, I just hate doing pointer math when it can be avoided

Reply:
Something feels a little off about taking an address of a vector item even though it is totally legit